### PR TITLE
r1.10.0 MegaMolBART Compatibility

### PR DESCRIPTION
**Pull request by michalivne**
# What does this PR do ?

1. Fixes RegExTokenizer missing attributes.
2. Fixes broken Megatron encoder-decoder LM model ability to receive directly encoder hiddens.

```
    MegatronLMEncoderDecoderModel.forward(
        self,
        encoder_input_ids,
        decoder_input_ids,
        encoder_attn_mask,
        decoder_attn_mask,
        token_type_ids=None,
        lm_labels=None,
        # NEW API
        enc_output=None,
        enc_output_attn_mask=None,
        # NEW API
        output_enc_hidden_only=False,
        enc_input=None,
    ):
```

3. Added MegatronLMEncoderDecoderModel.encode method.
4. Improved decoding speed.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -424,25 +431,95 @@ def loss_func(output_tensor):
...
+        # collect all arguments
+        all_args_name = args_name[:]
+        all_args = args[:]
+        for k, v in kwargs.items():
+            all_args_name.append(k)
+            all_args.append(v)
+
+        args_idx = [kwargs_to_arg_idx[n] for n in all_args_name]
+
+        # print(f"all_args_name = {all_args_name}   args_idx = {args_idx}")
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:474]: Clean up comment again maybe?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 2]
```diff
@@ -451,11 +528,87 @@ def id_func(output_tensor):
...
+
     def validation_step(self, batch, batch_idx):
+        """
+        return_values - if given, returns a dictionary with given keys and corresponding values
+        """
         batch_for_pipeline = self.process_global_batch(batch)
         encoder_seq_length = batch_for_pipeline[0].size(1)
         decoder_seq_length = batch_for_pipeline[1].size(1)
 
+        # # FIXME: finish this
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:595]: Does this need to be finished?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -743,7 +896,128 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
...
 
-    def decode(self, tokens_enc, enc_mask, num_tokens_to_generate, encoder_input=None, tokenizer=None):
+    def encode(self, tokens_enc, enc_mask, encoder_input=None, reconfigure_microbatch=True):
+        """
+        tokens_enc - encoder input tokens 
+        enc_mask - corresponding mask
+        encoder_input - encoder input (bypass tokens), if given tokens_enc can be None.
+        """
+        # Check whether the DDP is initialized. This is needed when running inference outside of training loop.
+        if parallel_state.is_unitialized():
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:906]: Can we move 906 - 945 to a common function that is called by encode() and decode()

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 4]
```diff
@@ -305,34 +305,44 @@ def set_input_tensor(self, input_tensor):
...
-                enc_input=enc_input, enc_attn_mask=enc_attn_mask, enc_layer_past=None, enc_get_key_value=False,
-            )
+            # When pipeline parallel > 1 we need to make sure encoder exist (will be missing in decoder)
+            if enc_output is None and self.enc_dec_model.encoder is not None:
+                enc_output = self.enc_dec_model.encode(
+                    enc_input=enc_input, enc_attn_mask=enc_attn_mask, enc_layer_past=None, enc_get_key_value=False,
+                )
+            else:
+                # # if pipeline parallel > 1 during encode decoder will return None which will create an error (so return 0 on correct device instead)
+                # t = enc_input if enc_input is not None else enc_output
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:336]: Can we remove comments again?

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Great PR! :) Left some comments mostly about cleaning up comments

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 5]
```diff
@@ -424,25 +431,95 @@ def loss_func(output_tensor):
...
+        return fwd_output_only_func
+
+    def get_forward_output_only_func_decode(self, arg_names, **kwargs):
+        """
+        args_idx - maps batch into index of args (with None filling gaps)
+        arg_names - corresponding names for a friendly error message
+        kwargs - shared arguments (non tensors)
+        """
+
+        def fwd_output_only_func(batch, model):
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:514]: Share with get_forward_output_only_func_encode, and add to docstring what are the outputs in id_func (e.g., 'logits')

**Review by michalivne — COMMENTED** [review]
Minor comments

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 5 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:532]: Not identical to decode. Let's discuss.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 3 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:924]: Not identical to decode. Let's discuss.

**Comment by michalivne** [comment]
Moved PR to fork to resolve branch protection..

**Comment by michalivne** [comment]
Moved PR to https://github.com/NVIDIA/NeMo/pull/4603 to resolve branch protection in main repo.
